packages/json-api/src/-private/cache.ts

```diff
@@ -525,38 +526,52 @@ export default class JSONAPICache implements Cache {
     const isUpdate = /*#__NOINLINE__*/ !_isEmpty(peeked) && !isLoading;
 
     if (LOG_OPERATIONS) {
+      logGroup(
+        'cache',
+        'upsert',
+        identifier.type,
+        identifier.lid,
+        existed ? 'merged' : 'inserted',
+        calculateChanges ? 'has-subscription' : ''
+      );
       try {
         const _data = JSON.parse(JSON.stringify(data)) as object;
+
         // eslint-disable-next-line no-console
-        console.log(`EmberData | Operation - upsert (${existed ? 'merge' : 'insert'})`, _data);
+        console.log(_data);
       } catch {
         // eslint-disable-next-line no-console
-        console.log(`EmberData | Operation - upsert (${existed ? 'merge' : 'insert'})`, data);
+        console.log(data);
       }
+      // eslint-disable-next-line no-console
+      console.groupEnd();
     }
 
     if (cached.isNew) {
       cached.isNew = false;
-      this._capabilities.notifyChange(identifier, 'identity');
-      this._capabilities.notifyChange(identifier, 'state');
+      this._capabilities.notifyChange(identifier, 'identity', null);
+      this._capabilities.notifyChange(identifier, 'state', null);
     }
 
-    if (calculateChanges) {
-      changedKeys = existed ? calculateChangedKeys(cached, data.attributes) : Object.keys(data.attributes || {});
+    // if no cache entry existed, no record exists / property has been accessed
+    // and thus we do not need to notify changes to any properties.
+    if (calculateChanges && existed && data.attributes) {
```

Comment:
this turns out to not be true for "emergent behaviors" in prod when belongsTo relationships are illegally accessed before load.

we fixed this by making the consuming record pay the cost for this by subscribing to its own "added" event and invalidating its attrs when so.